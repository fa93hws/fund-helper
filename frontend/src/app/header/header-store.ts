import { action, observable } from 'mobx';

export class HeaderStore {
  constructor(private readonly fetchValues: (id: string) => Promise<void>) {}

  @observable.ref
  idInput = '';

  @action
  setIdInput(input: string) {
    this.idInput = input;
  }

  @action
  async onSubmit() {
    await this.fetchValues(this.idInput);
  }
}
