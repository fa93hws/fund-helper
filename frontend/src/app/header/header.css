.bar {
  padding: 0 40px;
}

.subjectMatterInfo {
  margin-left: auto;
}
