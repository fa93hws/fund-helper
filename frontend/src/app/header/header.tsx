import * as React from 'react';
import { observer } from 'mobx-react';
import { IComputedValue } from 'mobx';
import AppBar from '@material-ui/core/AppBar';
import InputAdornment from '@material-ui/core/InputAdornment';
import InputBase from '@material-ui/core/InputBase';
import SearchIcon from '@material-ui/icons/Search';
import ToolBar from '@material-ui/core/Toolbar';
import Typography from '@material-ui/core/Typography';
import type {
  SubjectMatter,
  SubjectMatterInfo,
} from '../../services/subject-matter/subject-matter';
import { HeaderStore } from './header-store';
import styles from './header.css';

type HeaderProps = {
  idInput: string;
  info: SubjectMatterInfo | undefined;
  onIdChange: (id: string) => void;
  onSubmit: () => void;
};

export const Header = React.memo(
  ({ idInput, info, onSubmit, onIdChange }: HeaderProps) => {
    const handleSubmit = React.useCallback(
      (e: React.FormEvent<HTMLFormElement>) => {
        e.preventDefault();
        onSubmit();
      },
      [onSubmit],
    );
    const handleIdChange = React.useCallback(
      (e: React.ChangeEvent<HTMLInputElement>) => {
        onIdChange(e.target.value);
      },
      [onIdChange],
    );

    return (
      <AppBar position="static" className={styles.bar}>
        <ToolBar>
          <form onSubmit={handleSubmit}>
            <InputBase
              onChange={handleIdChange}
              value={idInput}
              style={{ color: 'white' }}
              margin="dense"
              placeholder="基金ID"
              startAdornment={
                <InputAdornment position="start">
                  <SearchIcon />
                </InputAdornment>
              }
            />
          </form>
          <div className={styles.subjectMatterInfo}>
            {info && (
              <Typography variant="h6" component="h6">
                {info.name}@{info.id} ({info.type})
              </Typography>
            )}
          </div>
        </ToolBar>
      </AppBar>
    );
  },
);

export function createHeader(
  fetchValues: (id: string) => Promise<void>,
  subjectMatter: IComputedValue<SubjectMatter | undefined>,
) {
  const store = new HeaderStore(fetchValues);
  const onIdChange = (id: string) => store.setIdInput(id);
  const onSubmit = () => store.onSubmit();
  return observer(() => (
    <Header
      idInput={store.idInput}
      info={subjectMatter.get()?.info}
      onSubmit={onSubmit}
      onIdChange={onIdChange}
    />
  ));
}
