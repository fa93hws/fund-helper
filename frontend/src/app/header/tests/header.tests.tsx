import * as React from 'react';
import { mount, render } from 'enzyme';
import { Header, FundInfo } from '../header';

describe('Header', () => {
  it('renders header without fund info and input', () => {
    expect(
      <Header
        idInput=""
        info={undefined}
        onIdChange={jest.fn()}
        onSubmit={jest.fn()}
      />,
    ).toMatchRenderedSnapshot();
  });

  it('renders header with fund info and input', () => {
    const info: FundInfo = {
      id: 'hanasaki',
      name: 'guming',
      typ: 'crystal',
    };
    expect(
      render(
        <Header
          idInput="cwx"
          info={info}
          onIdChange={jest.fn()}
          onSubmit={jest.fn()}
        />,
      ),
    ).toMatchSnapshot();
  });

  it('triggers onchange when text change', () => {
    const onChange = jest.fn();
    const component = mount(
      <Header
        idInput=""
        info={undefined}
        onIdChange={onChange}
        onSubmit={jest.fn()}
      />,
    );
    component.find('input').simulate('change');
    expect(onChange).toHaveBeenCalled();
  });
});
