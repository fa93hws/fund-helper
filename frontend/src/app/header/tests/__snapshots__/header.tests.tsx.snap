// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header renders header with subjectMatter info and input 1`] = `
<header
  class="MuiPaper-root MuiAppBar-root MuiAppBar-positionStatic MuiAppBar-colorPrimary MuiPaper-elevation4"
>
  <div
    class="MuiToolbar-root MuiToolbar-regular MuiToolbar-gutters"
  >
    <form>
      <div
        class="MuiInputBase-root MuiInputBase-adornedStart MuiInputBase-marginDense"
        style="color:white"
      >
        <div
          class="MuiInputAdornment-root MuiInputAdornment-positionStart"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M15.5 14h-.79l-.28-.27C15.41 12.59 16 11.11 16 9.5 16 5.91 13.09 3 9.5 3S3 5.91 3 9.5 5.91 16 9.5 16c1.61 0 3.09-.59 4.23-1.57l.27.28v.79l5 4.99L20.49 19l-4.99-5zm-6 0C7.01 14 5 11.99 5 9.5S7.01 5 9.5 5 14 7.01 14 9.5 11.99 14 9.5 14z"
            />
          </svg>
        </div>
        <input
          class="MuiInputBase-input MuiInputBase-inputAdornedStart MuiInputBase-inputMarginDense"
          placeholder="基金ID"
          type="text"
          value="cwx"
        />
      </div>
    </form>
    <div>
      <h6
        class="MuiTypography-root MuiTypography-h6"
      >
        guming@hanasaki (crystal)
      </h6>
    </div>
  </div>
</header>
`;

exports[`Header renders header without subjectMatter info and input 1`] = `
<header
  class="MuiPaper-root MuiAppBar-root MuiAppBar-positionStatic MuiAppBar-colorPrimary MuiPaper-elevation4"
>
  <div
    class="MuiToolbar-root MuiToolbar-regular MuiToolbar-gutters"
  >
    <form>
      <div
        class="MuiInputBase-root MuiInputBase-adornedStart MuiInputBase-marginDense"
        style="color:white"
      >
        <div
          class="MuiInputAdornment-root MuiInputAdornment-positionStart"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M15.5 14h-.79l-.28-.27C15.41 12.59 16 11.11 16 9.5 16 5.91 13.09 3 9.5 3S3 5.91 3 9.5 5.91 16 9.5 16c1.61 0 3.09-.59 4.23-1.57l.27.28v.79l5 4.99L20.49 19l-4.99-5zm-6 0C7.01 14 5 11.99 5 9.5S7.01 5 9.5 5 14 7.01 14 9.5 11.99 14 9.5 14z"
            />
          </svg>
        </div>
        <input
          class="MuiInputBase-input MuiInputBase-inputAdornedStart MuiInputBase-inputMarginDense"
          placeholder="基金ID"
          type="text"
          value=""
        />
      </div>
    </form>
    <div />
  </div>
</header>
`;
